### ISSUE TITLE
ARN No longer sent?

### ISSUE THREAD

**Issue by cristianoccazinsp** [bugsnag/bugsnag-react-native#376]
### Description

Somewhere between version updates (can't figure out which, tried a few older ones and no luck) and app name change, ARNs stopped getting sent. 

### Issue

I can't find out exactly if it was a version update or the fact that I've renamed the app, but Android's ARN are no longer getting sent. I've tried changing the manifest to set the ARN setting to true, but same thing.
Regular .notify calls are working, but native errors seems to have stopped working and are no longer getting sent. API KEY looks fine as well... Not sure where to start.

### Environment

Library versions:

```
"react-native": "0.59.9",
"react": "16.8.6",
"bugsnag-react-native": "2.21.1",


bugsnagConf.releaseStage = conf.env;
bugsnagConf.appVersion = `${versionText} (${conf.env})`;
bugsnagConf.autoCaptureSessions = conf.env != 'dev' ? true : false; 
bugsnagConf.automaticallyCollectBreadcrumbs = true;
bugsnagConf.autoNotify = conf.env != 'dev';
```

- iOS/Android version(s): 8 and 9
- simulator/emulator or physical device?: Physical (Moto G5 / Pixel 2)
- debug mode or production?: Both

- [x] (Android only) `BugsnagReactNative.start(this)` is present in the
  `onCreate` method of your `MainApplication` class? 
Yes

**Comment by mattdyoung** [comment]
Hi @cristianoccazinsp 

I assume you're referring to ANRs (Application Not Responding events on Android).

Bugsnag first added support for ANRs in bugsnag-react-native v2.16.0 (detection initially turned on by default), but due to customer feedback around the increased volume of events we set ANR detection off by default from v2.19.1 onwards. We also improved the detection algorithm used in v2.21.0 to remove false positives which were being reported on a few devices from specific manufacturers.

You say you're not receiving native exceptions or ANRs on Android?

I note above you've set `bugsnagConf.autoNotify = conf.env != 'dev'`, which would mean exceptions wouldn't be automatically in dev. I assume you're not testing this in dev?

To report Android ANRs using our React Native integration, you'll need to follow the enhanced native integration instructions for Android here:
https://docs.bugsnag.com/platforms/react-native/react-native/enhanced-native-integration/

And also ensure that you've set the `DETECT_ANRS` option in the `AndroidManifest.xml`:
https://docs.bugsnag.com/platforms/android/#reporting-anrs

If you still can't get this to work can you cut this down to an example project which reproduces the issue and share with us so we can see what might be wrong in the configuration of Bugsnag? You can email support@bugsnag.com if you'd rather not share the source code here.

Thanks!

**Comment by cristianoccazinsp** [comment]
Hello, thanks for the detailed response.

I believe I tried to set the manifest config on and didn't get ANRs. I will
try again just in case.

I have a very specific bug with react native camera that let's me trigger
ANRs pretty much on demand while recording videos. Any chance the updated
algorithm is no longer detecting this very specific case?

**Comment by mattdyoung** [comment]
The updated ANR detection algorithm uses a signal handler to detect `SIGQUIT` events so should be reliable and we're not aware of any issue with ANR detection in your scenario.

If you are able to reliably reproduce this though then if you can share a reproduction example project with us we can investigate to confirm the cause.

**Comment by cristianoccazinsp** [comment]
Even with ```<meta-data android:name="com.bugsnag.android.DETECT_ANRS" android:value="true"/>```, I'm still not seeing ANRs being reported.

I don't really have any easy to reproduce ANR, but basically while testing on a google pixel 2 (Android 9), the react-native-camera library hangs while recording video if the device rotation changes, or the device is locked. After this the app just shows the close app prompt.

I can see the ANR reported on the google console, however.

**Comment by mattdyoung** [comment]
I've just realised that you're initializing bugsnag-android via a `Configuration` object.

That will be overriding any values you set in the manifest so can you try enabling ANR detection in the `Configuration` object you pass to `Bugsnag.init` as shown here:
https://docs.bugsnag.com/platforms/android/configuration-options/#setdetectanrs

**Comment by cristianoccazinsp** [comment]
Just tried setting ```bugsnagConf.detectAnrs = true;``` as well, but no luck. I'm starting to think that this camera ANR is so odd that it is not even detected.

**Comment by mattdyoung** [comment]
Hi @cristianoccazinsp 

We'll discuss this amongst the team and investigate what could be causing these camera ANRs to be handled differently and not picked up by our detection mechanism.

I'll let you know once we know more.

**Comment by mattdyoung** [comment]
Hi @cristianoccazinsp 

Could you try one more thing? We believe if you're setting `detectAnrs` in your Javascript this won't be picked up in time by the native client in order to enable ANR detection.

In the `onCreate` method of your `MainApplication.java` can you try initializing the Bugsnag client like this:
```
// MainApplication class
@Override
public void onCreate() {
    super.onCreate();
    Configuration config = new Configuration("your-api-key-here");
    config.setDetectAnrs(true);
    BugsnagReactNative.startWithConfiguration(this /* app context */, config);    
    SoLoader.init(this, /* native exopackage */ false);
}
```

If that still doesn't work then this is likely to be something specific to the nature of the ANR you're triggering. If that's the case could you a cut down minimal project which demonstrates the issue so we can reproduce and try to resolve.

**Comment by cristianoccazinsp** [comment]
Hey @mattdyoung . That seems to work just fine! However, why is not picking the settings from the manifest file? I would really like to avoid having to configure everything in two locations.

**Comment by mattdyoung** [comment]
I agree, It's not ideal at the moment. We're currently looking at improving the way bugsnag-react-native and other error reporting libraries are configured so that more settings will be configurable up front in settings files and certain config options will be immutable rather than being able to redefine them later in code.

So the way this needs to be configured should improve in the near future.

**Comment by cristianoccazinsp** [comment]
Thank you, this was very useful!.

I'm wondering, do I still need to set the API key on the manifest file? Or now that the key is set directly in the java code it is no longer needed?

**Comment by mattdyoung** [comment]
Any options specified in the Configuration object in Java can be removed from the manifest.

Please be aware if you're also using our `bugsnag-android-gradle-plugin`, that also reads the API key from the manifest so in that case you'd need to specify the API key in the gradle plugin configuration, or continue to specify it in the manifest:
https://docs.bugsnag.com/build-integrations/gradle/#basic-configuration

**Comment by shercoder** [comment]
> SoLoader.init(this, /* native exopackage */ false);

@mattdyoung Do we have to match the configurations declared in JS and Java code or does the Java code just need to configure ANR?

**Comment by shercoder** [comment]
> Hi @cristianoccazinsp
> 
> Could you try one more thing? We believe if you're setting `detectAnrs` in your Javascript this won't be picked up in time by the native client in order to enable ANR detection.
> 
> In the `onCreate` method of your `MainApplication.java` can you try initializing the Bugsnag client like this:
> 
> ```
> // MainApplication class
> @Override
> public void onCreate() {
>     super.onCreate();
>     Configuration config = new Configuration("your-api-key-here");
>     config.setDetectAnrs(true);
>     BugsnagReactNative.startWithConfiguration(this /* app context */, config);    
>     SoLoader.init(this, /* native exopackage */ false);
> }
> ```
> 
> If that still doesn't work then this is likely to be something specific to the nature of the ANR you're triggering. If that's the case could you a cut down minimal project which demonstrates the issue so we can reproduce and try to resolve.

@mattdyoung Do we have to match the configurations declared in JS and Java code or does the Java code just need to configure ANR?

**Comment by mattdyoung** [comment]
@shercoder 

Several options (such as ANRs) are Android specific and can only be initialized in the Java code.

As our React Native library initializes, the native Android layer initializes first based on the Java configuration and then any options specified later as the JS layer initializes are passed through to the native layer.

This means that any native errors which occur early during app launch before the Javascript layer has initialized will only use the Java config, so you'll want to ensure any config affecting such native errors is specified there.

We're currently working on a project to simplify our React Native library so when that is released in the coming months this should be much simpler to configure.

**Comment by shercoder** [comment]
@mattdyoung Thank you for following up. And thanks for the update about the upcoming changes! Really looking forward to those!

Yeah, soon after I posted that comment, I actually dug into the library and found out exactly what you described above. :) It was nice to know the inners of the library. I was able to get it working.

One thing I'm not sure and I don't see any documentation is what is `SoLoader.init` for?
